### PR TITLE
Fix rmrf and drop version to 2.9.0

### PR DESCRIPTION
**Pull request by damccorm**
Current published version is 2.8.0. Probably not going to major version at this time.

### CODE REVIEW

**Comment by stephenmichaelf** [comment]
Can we add more details to this? What are we fixing? What's the new implementation? How is it different? Good to have for historical context.

**Comment by stephenmichaelf** [comment]
Are there tests that cover this? Do we need more?

**Comment by damccorm** [comment]
Sure, for context we're trying to solve the issue [here](https://github.com/Microsoft/azure-pipelines-tasks/issues/5472). To summarize, nodes unlinkSync doesn't always actually delete a file, it just decrements its ref-count (which eventually has the effect of deleting it). This means that if the file is open elsewhere (often with antivirus software its being scanned) it doesn't actually get immediately deleted. This tries to address that.

I think existing coverage does a pretty good job with this, not really sure if there's a good way to test that exact behavior. Since we're not using unlinkSync, that edge case should just be avoided entirely anyways.

File: node/task.ts [hunk 1]
```diff
@@ -1150,42 +1151,35 @@ function _legacyFindFiles_getMatchingItems(
...
-        throw new Error(loc('LIB_OperationFailed', 'rmRF', err.message));
     }
-
-    if (lstats.isDirectory()) {
-        debug('removing directory');
-        shell.rm('-rf', path);
-        let errMsg: string = shell.error();
-        if (errMsg) {
-            throw new Error(loc('LIB_OperationFailed', 'rmRF', errMsg));
+    catch (err) {
```
Comment (stephenmichaelf) [inline node/task.ts]: Did we lose the return? Hard to read the diff.

**Comment by stephenmichaelf** [comment]
Looks good, let's have Eric review before merging.

File: node/task.ts (hunk 1 above)
Comment (damccorm) [reply node/task.ts]: Yeah, we don't actually want that anymore though, we want to try to get rid of the dead symlink regardless

File: node/task.ts [hunk 2]
```diff
@@ -1150,42 +1151,35 @@ function _legacyFindFiles_getMatchingItems(
...
+        if (getPlatform() == Platform.Windows) {
+            if (fs.statSync(inputPath).isDirectory()) {
+                childProcess.execSync(`rd /s /q "${inputPath}"`);
+            }
+            else {
+                childProcess.execSync(`del /f /a "${inputPath}"`);
+            }
+        }
+        else {
+            childProcess.execSync('rm -rf ' + inputPath);
```
Comment (damccorm) [inline node/task.ts]: Use previous implementation for Linux/mac

File: node/task.ts [hunk 3]
```diff
@@ -1150,42 +1151,35 @@ function _legacyFindFiles_getMatchingItems(
...
         }
-
-        return;
     }
-
-    debug('removing file');
     try {
-        fs.unlinkSync(path);
+        // Still try to unlink in case its a dead symlink.
+        fs.unlinkSync(inputPath);
```
Comment (damccorm) [inline node/task.ts]: Don't need this now

File: node/task.ts [hunk 4]
```diff
@@ -1150,42 +1151,77 @@ function _legacyFindFiles_getMatchingItems(
...
-        // if you try to delete a file that doesn't exist, desired result is achieved
-        // other errors are valid
-        if (err.code == 'ENOENT') {
-            return;
+    if (getPlatform() == Platform.Windows) {
+        // Node doesn't provide a delete operation, only an unlink function. This means that if the file is being used by another
+        // program (e.g. antivirus), it won't be deleted. To address this, we shell out the work to rd/del.
+        try {
+            if (fs.statSync(inputPath).isDirectory()) {
+                debug('removing directory');
```
Comment (jtpetty) [inline node/task.ts]: is there a reason to not put the path you are trying to delete in the debug message?

Comment (damccorm) [reply node/task.ts]: Nope, will update

**Comment by damccorm** [comment]
I'm intentionally not updating the version in package.json. We should probably release all the changes going in and need to follow up in the release notes when we do.
